### ISSUE TITLE
Configuring Stream.Read(byte[], int, int) does not work - complains of mismatched out/ref args

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#508]
From [Richard Lang's Stackoverflow Question "A.Fake<Stream>().Read(…) throwing InvalidOperationException"](http://stackoverflow.com/questions/31419162/a-fakestream-read-throwing-invalidoperationexception):

When running 

```csharp
[Test]
public void test_whatIsUpWithStreamRead()
{
    Stream fakeStream = A.Fake<Stream>();

    A.CallTo(() => fakeStream.Read(A<byte[]>.Ignored, A<int>.Ignored, A<int>.Ignored))
        .Returns(1);

    byte[] buffer = new byte[16];

    int numBytesRead = fakeStream.Read(buffer, 0, 16);

    Assert.AreEqual(1, numBytesRead);
}
```

The `fakeStream.Read` 
throws a System.InvalidOperationException with the message:

> The number of values for out and ref parameters specified does not match the number of out and ref parameters in the call.

I've reproduced, and started debugging, but do not yet know where we're going wrong.

**Comment by blairconrad** [comment]
Update: I haven't seen the cause, and this is hasty, but FakeItEasy appears to think that we've supplied an our or ref value, but the faked method has no out or ref parameters (which it doesn't).
Seen while zipping in `BuildableCallRule.ApplyOutAndRefParametersValueProducer`. (I accidentally debugged too far, or I'd probably know how we got in that situation.)
Okay. Really going to the Day Job now. Will resume debugging tomorrow, if nobody else has solved it by then.

**Comment by blairconrad** [comment]
Let's ignore everything I said in the last comment. In fact, I'll go strike it out.

I noticed that when we parsed the `() => fakeStream.Read(A<byte[]>.Ignored, A<int>.Ignored, A<int>.Ignored)`, in `ExpressionArgumentConstraintFactory.GetArgumentConstraint`, we ended up thinking that the first parameter was an `out` parameter.
The processing FakeItEasy does on the parameter is slim to none, so I backed off a level and went right to the source, using reflection.
This test fails:

```csharp
[Test]
public void test_signature_of_read()
{
    var readMethod = typeof (Stream).GetMethods().Single(m => m.Name == "Read");

    var isReadMethodOut = readMethod.GetParameters()[0].IsOut;

    Assert.That(isReadMethodOut, Is.False);
}
```

Which explains why FakeItEasy is confused. I'm just not sure why the first parameter would register as `IsOut`.
Back to the Day Job, but if anyone has ideas, please pile 'em on.

**Comment by blairconrad** [comment]
I believe the array in `Stream.Read` is decorated with `[Out]`, which makes `ParameterInfo.IsOut` true, so `ExpressionArgumentConstraintFactory.GetArgumentConstraint` makes an `OutArgumentConstraint`.

I broke this in #343 (f7db76a8b3d5b78c3d9cd9f56a37a56d9656bea7) when trying to fix #215.
The change first made it into release 1.23.0.

My (new) understanding is if `IsByRef` is true, then the parameter is either an `out` or `ref`, with `IsOut` indicating that it's `out`.
The array parameter in `Stream.Read`'s `IsByRef` property is (predictably) false.

So `ExpressionArgumentConstraintFactory.IsOutArgument` should ensure not only that `IsOut` is true, but that the parameter type `IsByRef`.

**Comment by blairconrad** [comment]
@FakeItEasy/owners, I think I can whip up a fix fairly quickly.
Do we think it's worth releasing a 1.25.3? My preference would be to wait until 2.0.0, but progress on that release has stalled, so I've no feeling for how much longer that will be.

**Comment by blairconrad** [comment]
I am reconsidering my 2.0.0 comment. Forcing an upgrade to 2.0.0 in this case seems rude.

**Comment by blairconrad** [comment]
This issue has been fixed in release 1.25.3.

https://www.nuget.org/packages/FakeItEasy/1.25.3
https://github.com/FakeItEasy/FakeItEasy/releases/tag/1.25.3
